src/basic/mountpoint-util.c

```diff
@@ -182,12 +182,23 @@ int fd_is_mount_point(int fd, const char *filename, int flags) {
         int r;
 
         assert(fd >= 0);
-        assert(filename);
         assert((flags & ~(AT_SYMLINK_FOLLOW|AT_EMPTY_PATH)) == 0);
 
-        /* Insist that the specified filename is actually a filename, and not a path, i.e. some inode further
-         * up or down the tree then immediately below the specified directory fd. */
-        if (!filename_possibly_with_slash_suffix(filename))
+        /* Note that AT_EMPTY_PATH is only supported if the kernel supports statx(), or if the inode
+         * specified via 'fd' refers to a directory. Otherwise, we'll have to fail, because we have no kernel
+         * API to query the information we need.
+         *
+         * Also note that 'filename' being NULL will turn on AT_EMPTY_PATH implicitly. */
+
+        if (FLAGS_SET(flags, AT_EMPTY_PATH))
```

Comment:
Instead of all this normalization, why not just reduce the supported flags to only `AT_SYMLINK_FOLLOW` and infer `AT_EMPTY_PATH` solely from the `filename` argument?